OP_PAIRCOMMIT (#1699)
bitcoin/bips: bip-0442.md

    const valtype& vch2 = stacktop(-1);

    uint256 hash = PairCommitHash(vch1, vch2);

    stack.pop_back();
    stack.pop_back();
    stack.emplace_back(hash.begin(), hash.end());
    break;
}
```
```c++
const HashWriter HASHER_PAIRCOMMIT{TaggedHash("PairCommit")};

uint256 PairCommitHash(const std::vector<unsigned char>& x1, const std::vector<unsigned char>& x2)
{
    return (HashWriter{HASHER_PAIRCOMMIT} << x1 << x2).GetSHA256();
}
```

### Pull Request

https://github.com/lnhance/bitcoin/pull/6/files

## Rationale

### Cost comparison of Lightning Symmetry constructions

| Method           | ChannelSc | UpdateSc | UpdateW | ForceC  | Contest | Settle |
| :--------------- | --------: | -------: | ------: | ------: | ------: | :----: |
| APO-Annex        |      8 WU |   113 WU |  100 WU | 1221 WU |  627 WU | SigOp  |
| APO-Return       |      8 WU |   113 WU |   66 WU | 1359 WU |  765 WU | SigOp  |
| CTV+CSFS         |     43 WU |    81 WU |   98 WU | 1394 WU |  765 WU |  CTV   |
| CTV+CSFS+IKEY    |     10 WU |    48 WU |   98 WU | 1328 WU |  732 WU |  CTV   |
| CTV+CSFS+IKEY+PC |     11 WU |    49 WU |  131 WU | 1191 WU |  594 WU |  CTV   |

> *ChannelSc: channel script, UpdateSc: update script, UpdateW: witness size for Force Close and Contest, ForceC: cost of unilateral close, Contest: additional cost to contest, Settle: signature or CTV required for settlement.*

Author: Should add TEMPLATEHASH+IKEY+CSFS comparisions?

```suggestion
| Method           | ChannelSc | UpdateSc | UpdateW | ForceC  | Contest | Settle |
| :--------------- | --------: | -------: | ------: | ------: | ------: | :----: |
| APO-Annex        |      8 WU |   113 WU |  100 WU | 1221 WU |  627 WU | SigOp  |
| APO-Return       |      8 WU |   113 WU |   66 WU | 1359 WU |  765 WU | SigOp  |
| CTV+CSFS         |     43 WU |    81 WU |   98 WU | 1394 WU |  765 WU | HashEq |
| CTV+CSFS+IKEY    |     10 WU |    48 WU |   98 WU | 1328 WU |  732 WU | HashEq |
| CTV+CSFS+IKEY+PC |     11 WU |    49 WU |  131 WU | 1191 WU |  594 WU | HashEq |
| THIKCS-Annex     |     10 WU |    49 WU |  100 WU | 1160 WU |  563 WU | HashEq |
| THIKCS-Return    |     10 WU |    49 WU |   66 WU | 1295 WU |  699 WU | HashEq |

*ChannelSc: channel script, UpdateSc: update script, UpdateW: witness size for Force Close and Contest, ForceC: cost of unilateral close, Contest: additional cost to contest, Settle: signature or CTV required for settlement, THIKCS: TEMPLATEHASH+IKEY+CSFS.*
```

### Proving general computation using trees

> `OP_PAIRCOMMIT` enables Merkle tree commitments, which can represent functions and inputs/outputs. Taproot trees can be 128 levels deep, therefore including up to 2^128 possible leaf scripts. This is already over the practical limits to enumerate and Merklize.

Author: Is using 2<sup>128</sup> instead of 2^128 recommended?

```suggestion
`OP_PAIRCOMMIT` enables Merkle tree commitments, which can represent functions and inputs/outputs. Taproot trees can be 128 levels deep, therefore including up to 2<sup>128</sup> possible leaf scripts. This is already over the practical limits to enumerate and Merklize.
```

> `OP_PAIRCOMMIT` enables Merkle tree commitments, which can represent functions and inputs/outputs. Taproot trees can be 128 levels deep, therefore including up to 2<sup>128</sup> possible leaf scripts. This is already over the practical limits to enumerate and Merklize.

Author: This section seems a bit vague, how about?

```suggestion
`OP_PAIRCOMMIT` enables Merkle tree commitments, which can represent functions where inputs and corresponding outputs are the leaves. Taproot trees can be 128 levels deep, therefore including up to 2<sup>128</sup> possible leaf scripts that can represent solutions to complex functions. This is already over the practical computational limits to enumerate and Merklize.
```

<pre>
  BIP: 442
  Layer: Consensus (soft fork)
  Title: OP_PAIRCOMMIT
  Author: moonsettler <moonsettler@protonmail.com>
  Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-0442
  Status: Draft
  Type: Standards Track
  Created: 2024-12-09
  License: BSD-3-Clause
</pre>

## Abstract

This BIP describes a new tapscript opcode `OP_PAIRCOMMIT`, which
provides limited vector commitment functionality in tapscript.

When evaluated, the `OP_PAIRCOMMIT` instruction:
* Pops the top two values off the stack,

> * takes the "PairCommit" tagged SHA256 hash of the stack elements,
> * pushes the resulting commitment on the top of the stack.

Reviewer: ```suggestion
* takes the "PairCommit" tagged SHA256 hash of the stack elements' lengths and values,
* pushes the resulting 32-byte hash to the top of the stack.
```

## Motivation

To do [LN-Symmetry] contracts that don't require the nodes to keep old states,

> we need to solve the data availability problem presented by unilateral closes.

Reviewer: While this was the original motivation for developing PAIRCOMMIT, I don't think it's the current motivation for introducing it standalone, or including it in LNHANCE.

Perhaps something more like:

> When building scripts using `OP_CHECKSIGFROMSTACK`, it is common to require a commitment to multiple items in a specific sequence and without malleability. `OP_PAIRCOMMIT` enables efficient commitments of this type. Alternative methods for achieving these multiple commitments require an additional signature checking operation for each additional unordered commitment or 2 additional signature checks for each additional ordered commitment.
>
> One example of such a construction is [LN-Symmetry] which requires a single signature to commit both to the update CTV hash and the settlement CTV hash to ensure O(1) data storage requirements for each channel partner. Another is a complex delegation (delegating to various pubkeys after matching locktimes). Either of these contracts can be achieved without `OP_PAIRCOMMIT` but they would require the use of a costly `OP_RETURN` or 2 additional signature checks respectively.

Then (as we discussed) move the remaining symmetry discussion to its own section.

Channel peers must be able to reconstruct the script that spends an
intermediate state.

Using in sequence `OP_CHECKTEMPLATEVERIFY`, `OP_PAIRCOMMIT`, `OP_INTERNALKEY`
and `OP_CHECKSIGFROMSTACK` we can construct a [rebindable channel] that is also
[optimal].

The number of SHA256 iterations is minimized in the primary use case we
can optimize for, which is LN-Symmetry. Since the Tag can be pre-computed as
mid-state, it would only take 1 or 2 hash cycles in validation for the
unilateral close scenario.

## Specification

Repurpose opcode 205 (currently `OP_SUCCESS`) as follows:

`OP_PAIRCOMMIT` pops two elements off the stack, then concatenates them along
with their size commitments and takes the tagged SHA256 hash of that
concatenated string, then pushes the resulting hash back on the stack.

Given the stack `[x1, x2]`, where `x2` is at the top of the stack:

`OP_PAIRCOMMIT` will push `SHA256(tagPC|cs(x1)|x1|cs(x2)|x2)` onto the stack.

Where `|` denotes concatenation and `tagPC` is calculated according to
[BIP-340] tagged hash as `SHA256("PairCommit")|SHA256("PairCommit")` and

> `cs(x)` means `CompactSize(x)`.

Reviewer: Where is `CompactSize` defined?

Author: It's a function in bitcoin core and also used implicitly by the valtype serialization.

The following assembly-like pseudo-code shows a possible LN-Symmetry channel
construction that provides data availability to spend to the latest state from
an earlier state pushed on-chain with a forced close by channel partner.

```text
# S = 500000000
# IK -> A+B
<sig> <state-n-recovery-data> <state-n-hash> | CTV PC IK CSFS <S+1> CLTV DROP
```
before funding, sign the first state:
```text
# state-n-hash { nLockTime(S+n), out(contract, amount(A)+amount(B)) }
# settlement-n-hash { nSequence(2w), out(A, amount(A)), out(B, amount(B)) }
# state-n-recovery-data { settlement-n-hash or state-n-balance }

# contract for state n < m
IF
  <sig> <state-m-recovery-data> <state-m-hash> | CTV PC IK CSFS <S+n+1> CLTV DROP
ELSE
  <settlement-n-hash> CTV
ENDIF
```

### Use with future updates

Detailed introspection opcodes would also need vector commitments when used
with `OP_CHECKSIGFROMSTACK`.

`OP_CHECKCONTRACTVERIFY` would also need a way to carry complex data.

## Reference Implementation

A reference implementation is provided here:

https://github.com/lnhance/bitcoin/pull/6/files

## Rationale

If `OP_CAT` was available, it could be used to combine multiple stack elements

> that get verified with `OP_CHECKSIGFROMSTACK` as a valid state update.

Reviewer: Add
> Using `OP_CAT` for this purpose requires additional opcodes to prevent malleability (e.g. `0x0102 0x03 CAT` is identical to `0x01 0x0203 CAT`).

or similar.

> `OP_PAIRCOMMIT` solves this specific problem without introducing a wide range
> of potentially controversial new behaviors, such as novel 2-way peg mechanisms.

Reviewer: ```suggestion
`OP_PAIRCOMMIT` solves this specific problem without introducing granular introspection
via Andrew Poelstra's CAT and Schnorr Tricks.
```

Alternatively `OP_RETURN` could be used to ensure the availability of the state
recovery data, as `OP_CHECKTEMPLATEVERIFY` naturally commits to all outputs.
However, its cost in weight units would be over 4 times higher than that of
using `OP_PAIRCOMMIT`.

> 

Reviewer: Multiple data commitments can also be achieved using ephemeral key delegation to bind the data items.
```
witness: <asig> <a> <bsig> <b> <dsig> <dkey>
script: TUCK DUP TOALT TOALT <key> CHECKSIGFROMSTACK VERIFY \
    TUCK [check and consume b] FROMALT CHECKSIGFROMSTACK VERIFY \
    TUCK [check and consume a] FROMALT CHECKSIGFROMSTACK
```

One way to think about the 3 opcodes (`OP_CHECKSIGFROMSTACK`, `OP_INTERNALKEY`,
`OP_PAIRCOMMIT`) is we decompose a `OP_CHECKSIGFROMSTACK` variant that can use
a 1-byte `OP_TRUE` public key (substituting for the *taproot internal key*) and
can commit to a number of stack elements as a message.

> ### Behaviors LNhance tries to avoid introducing

Reviewer: Does this section belong in the PC BIP?

Author: Maybe we should have an LNhance BIP that depends on the others and move some stuff over?

Reviewer 2: I think it may be fine here if this section is moved up next to the Motivation section and the term LNHANCE is introduced clearly, ideally with a link to [lnhance.org](https://lnhance.org/) if that url is viable long-term.

* OP_VECTORCOMMIT: generalized form for n > 2 elements

### Cost comparison of LN-Symmetry constructions

| Method        | ChannelSc | UpdateSc | UpdateW | ForceC  | Contest | Settle |
| :------------ | --------: | -------: | ------: | ------: | ------: | :----: |
| APO-Annex     |      8 WU |   113 WU |  100 WU | 1221 WU |  627 WU | SigOp  |
| APO-Return    |      8 WU |   113 WU |   66 WU | 1359 WU |  765 WU | SigOp  |
| CTV+CSFS+IKEY |     10 WU |    48 WU |   98 WU | 1328 WU |  732 WU |  CTV   |
| CTV+CSFS      |     43 WU |    81 WU |   98 WU | 1394 WU |  765 WU |  CTV   |
| LNhance       |     11 WU |    49 WU |  131 WU | 1191 WU |  594 WU |  CTV   |

*ChannelSc: channel script, UpdateSc: update script, UpdateW: witness is the
same size for both Force Close and Contest in LN-Symmetry, ForceC: total cost of unilateral close transactions*

### Proving general computation

Merkle trees can be used to prove computation where the root of the tree
represents the *function* and the leaves represent the *inputs* and *output*.
There are practical limits to the entropy space for the *inputs* as they need
to be iterated over and hashed into a merkle root.

> MAST trees can currently cover 128 bits of entropy space, which is well over

Reviewer: MAST -> Taproot?

> the practical limits to iterate over and merklize. Therefore, we assume this

Reviewer: assume -> conclude

capability does not materially extend what computations are possible to prove
in bitcoin script. While `OP_PAIRCOMMIT` is not limited to a height of 128,

> that should not be practically feasible to utilize.

Reviewer: should not be -> is

Author: My confidence in that was not a 100% at the time. It was more of an open question posed to the dev community to falsify.

Reviewer 3: > should not be -> is

as in, it *is* practical to utilize deeper trees? (for this technique)

Did I miss a discovery?

I suppose a sparse tree could be deeper but that's logically equivalent to reducing the bit size of the inputs afaict.

Reviewer: Sorry, "is not"

> There is a way to reduce the size of the witness for proving computation,

Reviewer: I find this paragraph confusing. I think a variant of it probably belongs in the CHECKSIGFROMSTACK BIP - as CSFS can be used generally to sign data into a script and that is not limited to use with PC (although PC extends that capability to merkelized data).

Author: Thanks for the feedback, will take my time to go over it and consider every suggestion!

<pre>
  BIP: 442
  Layer: Consensus (soft fork)
  Title: OP_PAIRCOMMIT
  Author: moonsettler <moonsettler@protonmail.com>
  Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-0442
  Status: Draft
  Type: Standards Track
  Created: 2024-12-09
  License: BSD-3-Clause
</pre>

## Abstract

This BIP describes a new tapscript opcode `OP_PAIRCOMMIT`, which
provides limited vector commitment functionality in tapscript.

When evaluated, the `OP_PAIRCOMMIT` instruction:
* Pops the top two values off the stack,
* takes the "PairCommit" tagged SHA256 hash of the stack elements with size
commitments,
* pushes the resulting 32-byte hash to the top of stack.

## Motivation

> Currently, bitcoin lacks a way to hash multiple stack elements together. Which
> means building Merkle trees or verifying inclusion in a tree is not supported.

Reviewer: ```suggestion
Currently, bitcoin lacks a way to commit to multiple stack elements together. It is common practice to hash a single item as part of a bitcoin script, either in hash/time locked contracts, or in pay-to-pubkey-hash (P2PKH) scripts, but there is no way to commit to multiple items with a single hash. 

In a contrived but demonstrative example, if PAIRCOMMIT existed in legacy script, P2PKH could be extended to pay-to-pubkey-time-hash with scriptPubKey `2DUP PAIRCOMMIT RIPEMD160 <hash> EQUALVERIFY CHECKLOCKTIMEVERIFY DROP CHECKSIG`. This script format for single signature, time-locked bitcoin could be transformed into an address format.

With the ability to commit to pairs of elements, PAIRCOMMIT can be generalized Merklized commitments to a tree of elements with a single hash. On its own, this could enable a hash lock contract where the holder of any pre-image in a Merkle tree can unlock the spend, for example.

If `OP_CHECKSIGFROMSTACK` is combined with PAIRCOMMIT, the ability to sign commitments to multiple items enables both of the above but as a form of delegation. In cases where any single item can be used to unlock an output, additional off chain signatures be simply can be used as an alternative to PAIRCOMMIT. However in cases where multiple items must be used for spending together, PAIRCOMMIT removes the need for complex laddering schemes to ensure that all items were authorized together.
```

[BIP-340] tagged hash as `SHA256("PairCommit")|SHA256("PairCommit")` and
`cs(x)` means `CompactSize(x)`.

### Implementation

```c++
case OP_PAIRCOMMIT: {
    // OP_PAIRCOMMIT is only available in Tapscript
    // ...
    // x1 x2 -- hash
    if (stack.size() < 2) {
        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
    }
    const valtype& vch1 = stacktop(-2);
    const valtype& vch2 = stacktop(-1);

    uint256 hash = PairCommitHash(vch1, vch2);

    stack.pop_back();
    stack.pop_back();
    stack.emplace_back(hash.begin(), hash.end());
    break;
}
```
```c++
const HashWriter HASHER_PAIRCOMMIT{TaggedHash("PairCommit")};

uint256 PairCommitHash(const std::vector<unsigned char>& x1, const std::vector<unsigned char>& x2)
{
    return (HashWriter{HASHER_PAIRCOMMIT} << x1 << x2).GetSHA256();
}
```
### Use in script

`OP_PAIRCOMMIT` can be used to commit to a vector of stack elements in a way
that is not vulnerable to various forms of witness malleability. It is, however,
highly optimized for just 2 stack elements.

```text
# pc-hash = PC(a, PC(b, c))

<a> <b> <c> | PC PC <pc-hash> OP_EQUALVERIFY
```

> ### Use in LN-Symmetry

Reviewer 4: This is a subsection of the Specification section, but does "Use in LN-Symmetry" actually belong to the Specification? Wouldn’t it make more sense for it to be in the Rationale or a section with example uses?

To do LN-Symmetry contracts that don't require the nodes to keep old states,
we need to solve the data availability problem presented by unilateral closes.
Channel peers must be able to reconstruct the script that spends an
intermediate state.

> Using in sequence `OP_CHECKTEMPLATEVERIFY`, `OP_PAIRCOMMIT`, `OP_INTERNALKEY`

Reviewer 2: The pseudocode below might be clearer if the acronyms are defined beforehand

```suggestion
Using in sequence `OP_CHECKTEMPLATEVERIFY` (`CTV`), `OP_PAIRCOMMIT` (`PC`), `OP_INTERNALKEY` (`IK`)
```
(idem for CSFS on the next line)

and `OP_CHECKSIGFROMSTACK` we can construct a rebindable channel that is also
[optimal].

The following assembly-like pseudo-code shows a possible LN-Symmetry channel
construction that provides data availability to spend to the latest state from
an earlier state pushed on-chain with a forced close by channel partner.

```text
# S = 500000000
# IK -> A+B
<sig> <state-n-recovery-data> <state-n-hash> | CTV PC IK CSFS <S+1> CLTV DROP
```
before funding, sign the first state:
```text
# state-n-hash { nLockTime(S+n), out(contract, amount(A)+amount(B)) }
# settlement-n-hash { nSequence(2w), out(A, amount(A)), out(B, amount(B)) }
# state-n-recovery-data { settlement-n-hash or state-n-balance }

# contract for state n < m
IF
  <sig> <state-m-recovery-data> <state-m-hash> | CTV PC IK CSFS <S+n+1> CLTV DROP
ELSE
  <settlement-n-hash> CTV
ENDIF
```

> ### Use with future updates

Reviewer 4: Same as above, "Use with future updates" doesn’t seem to belong to the Specification.

Reviewer 2: Using `##` instead of `###` for this section and for "Use in LN-Symmetry" might suffice.

Detailed introspection opcodes would also need vector commitments when used

> with `OP_CHECKSIGFROMSTACK`.

Reviewer 2: nit, could link to or mention BIP348 with CHECKSIGFROMSTACK here.

> `OP_CHECKCONTRACTVERIFY` would also need a way to carry complex data.

Reviewer 2: This isn't part of any BIP yet, maybe add it later if that changes.

that get verified with `OP_CHECKSIGFROMSTACK` as a valid state update.

Using `OP_CAT` for this purpose requires additional opcodes to prevent witness
malleability (e.g. `0x0102 0x03 OP_CAT` is identical to `0x01 0x0203 OP_CAT`).

`OP_PAIRCOMMIT` solves this specific problem without introducing a wide range
of potentially controversial new behaviors like fully detailed introspection,
which includes the ability to inspect parent transactions and novel 2-way peg
mechanisms. ([CAT-tricks-I] and [CAT-tricks-II] by Andrew Poelstra)

Alternatively `OP_RETURN` could be used to ensure the availability of the state
recovery data, as `OP_CHECKTEMPLATEVERIFY` naturally commits to all outputs.
However, its cost in weight units would be over 4 times higher than that of
using `OP_PAIRCOMMIT`.

One way to think about the 3 opcodes (`OP_CHECKSIGFROMSTACK`, `OP_INTERNALKEY`,
`OP_PAIRCOMMIT`) is we decompose a `OP_CHECKSIGFROMSTACK` variant that can use
a 1-byte `OP_TRUE` public key (substituting for the *taproot internal key*) and
can commit to a number of stack elements as a message.

### Behaviors LNhance tries to avoid introducing

> The following behaviors are out of scope for LNhance and should not be enabled

Reviewer 4: This is the first mention of LNHANCE in this BIP, but is not further explained. LNHANCE is not introduced in this repository (except for a cursory mention in BIP 348 OP_CHECKSIGFROMSTACK). It would probably make sense to introduce the term, if this section is retained.

as a side effect without explicit consensus:

* Fine-grained introspection
* State-carrying covenants
* Bigint operations
* New arithmetic capabilities using lookup tables

### Alternative approaches

The following list of alternative approaches were discussed and rejected for
various reasons, either for expanding the scope or for unnecessary complexity:

* OP_CAT
* SHA256 streaming opcodes
* Merkle operation opcodes
* 'Kitty' CAT: result or inputs arbitrarily limited in size
* OP_CHECKTEMPLATEVERIFY committing to the taproot annex in tapscript
* OP_CHECKSIGFROMSTACK on n elements as message
* OP_VECTORCOMMIT: generalized form for n > 2 elements
* ReKey: key delegation and multiple use of OP_CHECKSIGFROMSTACK

> ### Cost comparison of LN-Symmetry constructions

Reviewer 4: ```suggestion
### Weight comparison of LN-Symmetry constructions
```

Reviewer 4: This comment was not addressed in the recent update.

Author: I see the problem with Cost when it comes to comparison of "fees", especially in regards to GSR adding actual execution cost to operations, but also don't like Weight so much because SigOps are in fact costly to validate.

> ### Cost comparison of LN-Symmetry constructions

Reviewer 2: It may make sense to put this section next to (or in) the "Use in LN-Symmetry" section.

> | Method        | ChannelSc | UpdateSc | UpdateW | ForceC  | Contest | Settle |
> | :------------ | --------: | -------: | ------: | ------: | ------: | :----: |
> | APO-Annex     |      8 WU |   113 WU |  100 WU | 1221 WU |  627 WU | SigOp  |
> | APO-Return    |      8 WU |   113 WU |   66 WU | 1359 WU |  765 WU | SigOp  |
> | CTV+CSFS+IKEY |     10 WU |    48 WU |   98 WU | 1328 WU |  732 WU |  CTV   |
> | CTV+CSFS      |     43 WU |    81 WU |   98 WU | 1394 WU |  765 WU |  CTV   |
> | LNhance       |     11 WU |    49 WU |  131 WU | 1191 WU |  594 WU |  CTV   |
> 
> *ChannelSc: channel script, UpdateSc: update script, UpdateW: witness is the
> same size for both Force Close and Contest in LN-Symmetry, ForceC: total cost of unilateral close transactions*

Reviewer 4: Some of these abbreviations are just a few characters shorter than the actual term. How about something along the lines of:

```suggestion
| Method        | Channel Script | Update Script | Update Witness¹ | Force Close²  | Contest | Settlement Mechanism |
| :------------ | --------: | -------: | ------: | ------: | ------: | :----: |
| APO-Annex     |      8 WU |   113 WU |  100 WU | 1221 WU |  627 WU | SigOp  |
| APO-Return    |      8 WU |   113 WU |   66 WU | 1359 WU |  765 WU | SigOp  |
| CTV+CSFS+IKEY |     10 WU |    48 WU |   98 WU | 1328 WU |  732 WU |  CTV   |
| CTV+CSFS      |     43 WU |    81 WU |   98 WU | 1394 WU |  765 WU |  CTV   |
| LNhance       |     11 WU |    49 WU |  131 WU | 1191 WU |  594 WU |  CTV   |

¹ *witness is the same weight for both Force Close and Contest in LN-Symmetry*  
² *total cost of unilateral close transactions*
```

Reviewer 4: This table is also provided barren of context. It would probably be useful to link to the compared schemas, or to the source of the table, and to add a couple sentence as to what we are looking at in the first place.

### Proving general computation

Merkle trees can be used to prove computation where the root of the tree
represents the *function* and the leaves represent the *inputs* and *output*.
There are practical limits to the entropy space for the *inputs* as they need
to be iterated over and hashed into a Merkle root.

> Taproot MAST trees can currently cover 128 bits of entropy space, which is over

Reviewer 4: What are "Taproot MAST trees"? The taproot BIPs speaks of "Script Trees" and MAST would be understood as "Merklized _Alternative Script_ Trees" in the context of Taproot. On the other hand, MAST as described in BIP 114 and BIP 117 speaks of Merklized _Abstract Syntax_ Trees where multiple leaves are used in combination.

You mention MAST here in the context taproot, where only a single leaf is ever used at once, but it seems to me that you might be thinking of Abstract Syntax Trees. Either way, it is not clear to me what you are referring to here, when you speak of the "entropy space covered by Taproot MAST trees" while treating leaves as "inputs and outputs".

This section is jumping right in the middle of something. I’m not sure that the scenario is described sufficiently to understand what the supposed concern is, let alone why the supposed concern is not a concern. If this section is relevant to this BIP, I reckon that more context is necessary for the reader to understand why and how, e.g., add aa link to a description of this scheme, or provide a bit more explanation what this section is talking about.

Reviewer 5: IMO this is a problem in the taproot BIP series, where various individuals decided it pertinent to change a well understood acronym (MAST) to suit taproot, rather than resepcting an existing body of literature including other BIPs.

I'd rather the taproot BIPs get updated to reflect the terminology Merklized Abstract Syntax Trees (or Tap Script Merkle Commitments, TSMC, if we're fond of confusing backronyming but should be less ambiguous in context) rather than Merklized Alternative Script Trees, which was a backronym.

Reviewer 5: correction: it seems there is no mention of alt script tree in the bips repos, just abstract syntax tree, including in the BIPs for Taproot itself. It's just a confusing rename people have made outside of this repo.

Author: So should it be "Taptrees" instead of "Taproot MAST trees"? Or "Tapscript trees"?

* takes the "PairCommit" tagged SHA256 hash of the stack elements with size commitments,
* pushes the resulting 32-byte hash to the top of stack.

## Specification

The notation below follows that of [BIP-340]. This includes the
$$hash_{tag}(x)$$ notation to refer to
$$SHA256(SHA256(tag) \\| SHA256(tag) \\| x)$$.

* If fewer than 2 elements are on the stack, the script MUST fail and
  terminate immediately.
* The top element ($$x2$$) and second to top element ($$x1$$) are read from the
  stack.
* Let $$pc$$ be $$hash_{PairCommit}(
  compact\\_size(size\\:of\\:x1) \\| x1 \\|
  compact\\_size(size\\:of\\:x2) \\| x2)$$[^1]
* The top two elements are popped from the stack.
* $$pc$$ is pushed to the stack.

[^1]: The number of SHA256 blocks is minimized in typical use cases. Since the
    Tag can be pre-computed as a SHA256 mid-state, it takes only 2 hash cycles
    for a commitment to 2 32-byte items or 1 for 2 smaller items.

## Motivation

Currently, bitcoin lacks a way to commit to multiple data items together. It
is common practice to hash a single item as part of a bitcoin script. Either
in hash/time locked contracts, or in pay-to-pubkey-hash (P2PKH) scripts, but

> there is no way to commit to multiple items with a single hash. 

Reviewer 4: The sentence starting with "Either" seems to be missing one or more words.

<a> <b> <c> | PC PC <pc-hash> OP_EQUALVERIFY
```

> ### Use in Lightning Symmetry

Reviewer 4: I think "LN-Symmetry" is more commonly used to refer to the concept than "Lightning Symmetry". Is there a specific reason why you prefer the latter?

To create Lightning Symmetry contracts that neither require the nodes to keep
old states nor require additional signing round trips (and corresponding
signature validations), we need to solve a data availability problem presented

> by contested closes. Specifically, the second to act channel partner must be
> able to reconstruct the script that spends the first update transaction
> published. This script includes the hash of the corresponding settlement

Reviewer 4: Maybe this is easier to parse:

```suggestion
by contested closes. Specifically, either channel partner must be
able to spend the output of any prior update transaction. 
Those input scripts include the hash of the corresponding settlement
```

}
```

### PR

https://github.com/lnhance/bitcoin/pull/6/files

## Rationale

### Cost comparison of Lightning Symmetry constructions

The following table briefly summarizes the costs associated with force closing
Lightning Symmetry channels enabled by various combinations of proposed script
upgrades.

| Method           | ChannelSc | UpdateSc | UpdateW | ForceC  | Contest | Settle |
| :--------------- | --------: | -------: | ------: | ------: | ------: | :----: |
| APO-Annex        |      8 WU |   113 WU |  100 WU | 1221 WU |  627 WU | SigOp  |
| APO-Return       |      8 WU |   113 WU |   66 WU | 1359 WU |  765 WU | SigOp  |
| CTV+CSFS         |     43 WU |    81 WU |   98 WU | 1394 WU |  765 WU |  CTV   |
| CTV+CSFS+IKEY    |     10 WU |    48 WU |   98 WU | 1328 WU |  732 WU |  CTV   |
| CTV+CSFS+IKEY+PC |     11 WU |    49 WU |  131 WU | 1191 WU |  594 WU |  CTV   |

*ChannelSc: channel script, UpdateSc: update script, UpdateW: witness is the
same size for both Force Close and Contest in Lightning Symmetry, ForceC: total
cost of unilateral close transactions, Contest: The additional cost to contest
a force closure, Settle: Whether a signature operation or CTV operation is

> required to validate the settlement transaction*

Reviewer 4: Thanks for expanding on the column titles. As I wrote above, it would be good to provide more context for this table. It would especially help if the compared schemata were at least briefly described and/or sources for details were provided.

### Proving general computation using trees

One potential risk of PAIRCOMMIT is that it may enable verification of general
computations on bitcoin.

Any script change which enables commitment to pairs of items necessarily
enables commitments to Merkle trees.

Merkle trees can be used to prove computation where the root of the tree
represents the *function* and the leaves represent the *inputs* and *output*.
There are practical limits to the total number of discrete output values that
can be reached by all possible input values (i.e. the input entropy space) as
they must be enumerated and hashed into a Merkle root.

Taproot trees can be 128 levels deep, therefore including up to 2^128 possible

> output values. This is over the practical limits to enumerate and merkelize.

Reviewer 4: Consistency nit:

```suggestion
output values. This is over the practical limits to enumerate and Merklize.
```

> output values. This is over the practical limits to enumerate and merkelize.

Reviewer 4: The first time I read this paragraph, I was confused and didn’t realize that you were comparing the expressional power of OP_PAIRCOMMIT and P2TR.
Perhaps it would help if you write "leaf scripts" instead of "output values" here:

```suggestion
leaf scripts. This is over the practical limits to enumerate and merkelize.
```

Therefore, we conclude that enabling Merkle commitments in script does not
materially extend what computations are possible to validate in bitcoin
script. While `OP_PAIRCOMMIT` is not explicitly limited to a height of 128, greater

> heights are not practical to Merkelize.

Reviewer 4: Now you have found three different ways to spell it. ;)
The researcher’s name is Ralph Merkle, so either "merklized" or "Merklized" seems right.

```suggestion
heights are not practical to Merklize.
```

Reviewer 2: We had a discussion about this a few years ago on the optech repo and IIRC Ralph Merkle himself replied in favor of "merklize"...

https://github.com/bitcoinops/bitcoinops.github.io/blob/master/STYLE.md#spelling

Reviewer 2: ah, here it is: https://github.com/bitcoinops/bitcoinops.github.io/issues/171#issuecomment-529167731
